### PR TITLE
Add test for TableExistsPrecondition

### PR DESCRIPTION
**Pull request by crizzis**
## Environment

**Liquibase Version**: 4.7.1

**Liquibase Integration & Version**: core

**Liquibase Extension(s) & Version**: none

**Database Vendor & Version**: Postgres 12

**Operating System Type & Version**: all (tested on Mac OS 10.15.7)

## Pull Request Type

- [x] Bug fix (non-breaking change which fixes an issue.)
- [ ] Enhancement/New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Added integration test showing that #1712 can no longer be reproduced

## Actual Behavior
- TableExistsPrecondition correctly identifies existing tables by name

## Expected/Desired Behavior
- TableExistsPrecondition correctly identifies existing tables by name

## Additional context

Depends on changes done [in this PR]https://github.com/liquibase/liquibase/pull/2529. Will rebase before merging. 

## Fast Track PR Acceptance Checklist:

- [x] Build is successful and all new and existing tests pass
- [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
- [x] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
- [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
- [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

### CODE REVIEW

**Comment by kataggart** [comment]
Thank you so much! We are reviewing.

**Review by nvoxland — APPROVED** [review]

**Comment by nvoxland** [comment]
Code Review/Testing notes:

- The only change in this PR is https://github.com/liquibase/liquibase/pull/2538/commits/f9bd8e571465f61ec5f79eff9cc9443170cd860b
  - The rest are from #2529, so this needs to be merged AFTER that PR is merged. The other commits will disappear from the "files" tab once that is merged.

File: liquibase-core/src/main/java/liquibase/util/SqlUtil.java [hunk 1]
```diff
@@ -120,7 +123,7 @@ public static Object parseValue(Database database, Object val, DataType type) {
         }
 
         String typeName = type.getTypeName();
-        try (Scanner scanner = new Scanner(stringVal.trim())) {
+        try (Scanner scanner = new Scanner(stringVal.trim()).useLocale(US)) {
```
Comment (XDelphiGrl) [inline liquibase-core/src/main/java/liquibase/util/SqlUtil.java:126]: @nvoxland, should we be hard-coding the locale in this try statement? If so, can you please explain why? Thank you!

File: liquibase-core/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.8.xsd [hunk 2]
```diff
@@ -0,0 +1,1400 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (XDelphiGrl) [inline liquibase-core/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.8.xsd:1]: @nvoxland, hi -Is it usual for us to update an older XSD version? Perhaps this is some strange merge issue?

File: liquibase-extension-testing/src/main/groovy/liquibase/extension/testing/command/CommandTests.groovy [hunk 3]
```diff
@@ -460,7 +460,6 @@ Long Description: ${commandDefinition.getLongDescription() ?: "NOT SET"}
                         throw new ComparisonFailure("$outputDescription does not contain expected", expectedOutputCheck, fullOutput)
                     }
                 } else if (expectedOutputCheck instanceof Pattern) {
-                    String patternString = StringUtil.standardizeLineEndings(StringUtil.trimToEmpty(((Pattern) expectedOutputCheck).pattern()))
```
Comment (XDelphiGrl) [inline liquibase-extension-testing/src/main/groovy/liquibase/extension/testing/command/CommandTests.groovy:463]: Why was this line removed, please?

Comment (crizzis) [reply liquibase-extension-testing/src/main/groovy/liquibase/extension/testing/command/CommandTests.groovy:463]: The result was never used

File: liquibase-core/src/main/java/liquibase/util/SqlUtil.java (hunk 1 above)
Comment (XDelphiGrl) [reply liquibase-core/src/main/java/liquibase/util/SqlUtil.java:126]: hi, @crizzis, thank you for the explanation. I didn't realize that `Scanner` was locale-sensitive. Were you on a machine with a different locale when you were coding the fix & saw a problem?

File: liquibase-extension-testing/src/main/groovy/liquibase/extension/testing/command/CommandTests.groovy (hunk 3 above)
Comment (XDelphiGrl) [reply liquibase-extension-testing/src/main/groovy/liquibase/extension/testing/command/CommandTests.groovy:463]: That's as good a reason as any to remove that line! Resolving this conversation.

**Comment by crizzis** [comment]
Yup. As a matter of fact the tests wouldn't pass for me

**Comment by XDelphiGrl** [comment]
> Yup. As a matter of fact the tests wouldn't pass for me

I thought that might be the case. :) Thanks for answering my questions! It's lovely to "meet" you!

**Comment by XDelphiGrl** [comment]
This PR adds a new integration test for the Liquibase tableExists precondition. The test is intended to validate that an existing table is correctly discovered during Liquibase update. 

**Special Thanks!**
@crizzis, thanks for adding tests! As a Software Quality Engineer, I have a special place in my heart for developers who write tests. :heart:

---

The bug in the issue linked to this PR is not fixed. I am able to reproduce the issue both on 4.8.0 and on the branch build for this PR. I believe that the test here might not be showing exactly what the bug is. That said, I find it difficult to understand the integration tests and perhaps have misunderstood. The new test does not hurt anything, and I am in favor of taking this PR. 

[See issue for more detailed test details.](https://github.com/liquibase/liquibase/issues/1712#issuecomment-1063147229)

@nvoxland, do you want to keep this PR open and add to it to address the other issue?

---
**Test Changelog**
```
databaseChangeLog:
- changeSet:
    id: GH1712-1
    author: liquibase
    preConditions:
      not:
        tableExists:
          tableName: gh1712_tbl
    changes:
    - createTable:
        columns:
          - column:
              constraints:
                nullable: false
                primaryKey: true
                primaryKeyName: gh1712_pk
              name: uuid
              type: UUID
        tableName: gh1712_tbl
        
        
- changeSet:
    id: GH1712-2
    author: liquibase
    preConditions:
      not:
        tableExists:
          tableName: gh1712a_tbl
    changes:
    - createTable:
        columns:
          - column:
              constraints:
                nullable: false
                primaryKey: true
                primaryKeyName: gh1712a_pk
              name: uuid
              type: UUID
        tableName: gh1712a_tbl
        
- changeSet:
    id: GH1712-3
    author: liquibase
    preConditions:
      onFail: WARN
      not:
        tableExists:
          tableName: gh1712a_tbl
    changes:
    - createTable:
        columns:
          - column:
              constraints:
                nullable: false
                primaryKey: true
                primaryKeyName: gh1712b_pk
              name: uuid
              type: UUID
        tableName: gh1712b_tbl
```
---
**Test Environment**
Liquibase Core: bugfix/1712/1822/513c90, Pro: master/691/1a5a8e
Postgres 12.6
[Passing Functional Tests](https://github.com/liquibase/liquibase-pro-tests/actions/runs/1959367329)

**Comment by nvoxland** [comment]
@XDelphiGrl I think your test changelog is showing that it is actually fixed in master.

When I run yours against master, I get 

```
[2022-03-10 11:53:00] WARNING [liquibase.precondition] Executing com/example/changelog.demo.yaml::GH1712-3::liquibase despite precondition failure due to onFail='WARN':
      1 preconditions failed
     com/example/changelog.demo.yaml : Not precondition failed
```

which has GH1712-3 failing because `gh1712a_tbl` _does_ exist since it was created in the changeset above. 

Or are you seeing something different as wrong?
